File: web/src/lib/activity.ts
import type { TodoItem, TodoList } from "./types";

/**
 * Verlauf einer Liste.
 *
 * Bewusst abgeleitet aus `doneBy` und `doneAt` der Todos statt aus einer
 * eigenen Collection: das kostet keine zusätzlichen Schreibvorgänge und
 * braucht keine Aufräumregel. Der Preis dafür ist, dass nur das Abhaken
 * auftaucht und immer nur der letzte Stand – wer abhakt und wieder aufmacht,
 * verschwindet aus dem Verlauf.
 *
 * Für echten Verlauf („Tom hat Brot hinzugefügt", „Jana hat den Preis
 * geändert") bräuchte es eine Subcollection und eine Cloud Function, die sie
 * füllt. Solange die einfache Variante reicht, lohnt das nicht.
 */
export interface ActivityEntry {
  todoId: string;
  title: string;
  /** Anzeigename; leer, wenn die Person nicht mehr in der Liste steht. */
  who: string;
  at: Date;
}

/** Wie viele Einträge der Verlauf höchstens zeigt. */
const LIMIT = 20;

export function activityOf(list: TodoList, todos: TodoItem[]): ActivityEntry[] {
  return todos
    .filter((todo) => todo.isDone && todo.doneAt !== null)
    .map((todo) => ({
      todoId: todo.id,
      title: todo.title,
      who: todo.doneBy ? (list.memberNames[todo.doneBy] ?? "") : "",
      at: todo.doneAt!.toDate(),
    }))
    .sort((a, b) => b.at.getTime() - a.at.getTime())
    .slice(0, LIMIT);
}

/**
 * „vor 5 Minuten", „gestern", „12.09.".
 *
 * Intl.RelativeTimeFormat übernimmt die Sprachformen; die Auswahl der Einheit
 * bleibt hier, weil die Bibliothek nur formatiert und nicht entscheidet.
 */
const relative = new Intl.RelativeTimeFormat("de-DE", { numeric: "auto" });
const shortDate = new Intl.DateTimeFormat("de-DE", { day: "2-digit", month: "2-digit" });

export function formatWhen(at: Date, now: Date = new Date()): string {
  const seconds = Math.round((at.getTime() - now.getTime()) / 1000);
  const absolute = Math.abs(seconds);

  if (absolute < 60) return "gerade eben";
  if (absolute < 3600) return relative.format(Math.round(seconds / 60), "minute");
  if (absolute < 86400) return relative.format(Math.round(seconds / 3600), "hour");
  // Ab einer Woche wird „vor 9 Tagen" unhandlich – dann lieber das Datum.
  if (absolute < 7 * 86400) return relative.format(Math.round(seconds / 86400), "day");
  return shortDate.format(at);
}

File: app/src/main/java/com/beigel/list2share/data/Activity.kt
package com.beigel.list2share.data

import android.text.format.DateUtils
import java.util.Date

/**
 * Verlauf einer Liste.
 *
 * Bewusst abgeleitet aus `doneBy` und `doneAt` der Todos statt aus einer
 * eigenen Collection: das kostet keine zusätzlichen Schreibvorgänge und
 * braucht keine Aufräumregel. Der Preis dafür ist, dass nur das Abhaken
 * auftaucht und immer nur der letzte Stand – wer abhakt und wieder aufmacht,
 * verschwindet aus dem Verlauf.
 *
 * Für echten Verlauf („Tom hat Brot hinzugefügt", „Jana hat den Preis
 * geändert") bräuchte es eine Subcollection und eine Cloud Function, die sie
 * füllt. Solange die einfache Variante reicht, lohnt das nicht.
 */
data class ActivityEntry(
    val todoId: String,
    val title: String,
    /** Anzeigename; leer, wenn die Person nicht mehr in der Liste steht. */
    val who: String,
    val at: Date,
)

/** Wie viele Einträge der Verlauf höchstens zeigt. */
private const val LIMIT = 20

fun activityOf(list: TodoList, todos: List<TodoItem>): List<ActivityEntry> =
    todos.asSequence()
        .filter { it.isDone && it.doneAt != null }
        .map { todo ->
            ActivityEntry(
                todoId = todo.id,
                title = todo.title,
                who = todo.doneBy?.let { list.memberNames[it] } ?: "",
                at = todo.doneAt!!.toDate(),
            )
        }
        .sortedByDescending { it.at.time }
        .take(LIMIT)
        .toList()

/**
 * „vor 5 Minuten", „gestern", „12.09.".
 *
 * DateUtils übernimmt Sprache und Einheit; ab einer Woche liefert es von
 * selbst ein Datum statt einer immer unhandlicheren Anzahl von Tagen.
 */
fun formatWhen(at: Date, now: Long = System.currentTimeMillis()): String =
    DateUtils.getRelativeTimeSpanString(
        at.time,
        now,
        DateUtils.MINUTE_IN_MILLIS,
        DateUtils.FORMAT_ABBREV_RELATIVE,
    ).toString()
